spring-cloud-kubernetes-fabric8-discovery/src/main/java/org/springframework/cloud/kubernetes/fabric8/discovery/KubernetesCatalogWatch.java

```diff
@@ -19,35 +19,34 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Objects;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 import io.fabric8.kubernetes.api.model.EndpointAddress;
 import io.fabric8.kubernetes.api.model.EndpointSubset;
 import io.fabric8.kubernetes.api.model.Endpoints;
 import io.fabric8.kubernetes.api.model.ObjectReference;
 import io.fabric8.kubernetes.client.KubernetesClient;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import org.apache.commons.logging.LogFactory;
 
 import org.springframework.cloud.client.discovery.event.HeartbeatEvent;
 import org.springframework.cloud.kubernetes.commons.discovery.KubernetesDiscoveryProperties;
 import org.springframework.context.ApplicationEventPublisher;
 import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.core.log.LogAccessor;
 import org.springframework.scheduling.annotation.Scheduled;
 
 /**
  * @author Oleg Vyukov
  */
 public class KubernetesCatalogWatch implements ApplicationEventPublisherAware {
 
-	private static final Logger logger = LoggerFactory.getLogger(KubernetesCatalogWatch.class);
+	private static final LogAccessor LOG = new LogAccessor(LogFactory.getLog(KubernetesCatalogWatch.class));
 
 	private final KubernetesClient kubernetesClient;
 
 	private final KubernetesDiscoveryProperties properties;
 
-	private final AtomicReference<List<String>> catalogEndpointsState = new AtomicReference<>();
```

Comment:
We were using `AtomicReference` here, but only methods `get/set` from it, and these are documented as : 

> Returns the current value, with memory effects as specified by VarHandle.getVolatile.

and : 

> Returns the value of a variable, with memory semantics of reading as if the variable was declared volatile.

So in essence, this is just `volatile` set and get. As such, we can simplify this to simply a `volatile` variable.

Reply:
To be fair, this is not enough either. 

A volatile _write_ (`catalogEndpointsState = endpointsPodNames`) is not guaranteed to be seen by the read _immediately_. In our case (if `@Scheduled` is performed by multiple threads, of course), it means that `List<String> previousState = catalogEndpointsState;` is not obligated to see the write to the volatile variable (`catalogEndpointsState = endpointsPodNames`) immediately; at least not according to the JLS specification. 

The specification says : 

> A write to a volatile field happens-before every "subsequent" read of that same field.

Subsequent here means "_when_ the read is observed". In practice though, a "cache invalidation" event is issued across all CPUs. In our case, I don't think this matters a lot. In theory, we can send a heartbeat with a list of endpoints that are not up-to-date (highly highly unlikely though), but we will send a heartbeat with all endpoints eventually.

Reply:
If we really want atomic updates (and be consistent with JLS), we should use proper semantics: the simplest form here, imho, would be a ReentrantLock, but that would mean locking... For the time being, I would leave it like this, and if we get reports, we can change. wdyt?